### PR TITLE
first commit of dev script

### PR DESCRIPTION
**Pull request by josieburgin**

### CODE REVIEW

File: bulkLoadDatasets.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        return False
+    datasetsInRegistry = response.json()
+    if "datasets" in datasetsInRegistry:
+        if datasetsInRegistry["datasets"]:
+            if includes(datasetsInRegistry, dataset):
+                return False
+            else:
+                return True
+        else:
+            return True
```
Comment (Alexey-ebi) [inline bulkLoadDatasets.py:100]: This could be rewritten as return not includes(datasetsInRegistry, dataset

File: bulkLoadDatasets.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    try:
+        url = MGX_GET.format(insdc_id)
+        response = requests.get(url)
+        return response
+    except Exception as e:
+        logging.error("Error accessing MGX registry: {0}".format(e))
+        return None
+
+
+def dataset_is_public(dataset):
```
Comment (Alexey-ebi) [inline bulkLoadDatasets.py:124]: Docstring is required; consider to use @retry decorator here

File: bulkLoadDatasets.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    if config.PUBLIC_CHECK_ENDPOINT != '':
```
Comment (Alexey-ebi) [inline bulkLoadDatasets.py:125]: could be simplified to just if config.PUBLIC_CHECK_ENDPOINT (empty string returns false in python anyway)

File: bulkLoadDatasets.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        if response.status_code == 200:
+            return True
+        else:
+            logging.info('Dataset is not public within source resource')
+            return False
+    else:
+        return True
+
+
+def post_dataset(dataset):
```
Comment (Alexey-ebi) [inline bulkLoadDatasets.py:137]: Docstring is required

File: bulkLoadDatasets.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    try:
+        logging.info(f'Posting dataset to registry: {dataset["sequenceID"]}')
+        apikey = 'mgx ' + config.AUTHORISATION_TOKEN
+        headers = {'Authorization': apikey}
+        r = requests.post(MGX_POST, json=dataset, headers=headers)
+        logging.info(f'Dataset not posted to registry: {r.content}')
+        time.sleep(1)
+    except Exception as e:
+        logging.error("Error posting dataset to MGX registry: {0}".format(e))
```
Comment (Alexey-ebi) [inline bulkLoadDatasets.py:146]: Consider to use @retry decorator here as previously (it has sleep option also)

File: bulkLoadDatasets.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+import sys
+import pandas
+import requests
+import logging
+
+import configmgrast as config
+
+
+def set_parser():
+    parser = argparse.ArgumentParser(prog='loadDatasets',
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py:22]: I've added docstrings to all the methods now

File: bulkLoadDatasets.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                                     description='Downloads INSDC ID to source ID mappings and loads datasets '
+                                                 'into the Metagenome Exchange Registry')
+    parser.add_argument('-v', '--version', action='version', version='%(prog)s 1.0.0')
+    return parser
+
+
+run_pattern = re.compile('^[EDS]RR[0-9]{6,7}$')
+source_pattern = re.compile(config.SOURCE_PATTERN)
+MGX_GET = 'https://wwwdev.ebi.ac.uk/ena/registry/metagenome/api/sequences/{}/datasets'
+MGX_POST = 'https://wwwdev.ebi.ac.uk/ena/registry/metagenome/api/admin/datasets'
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py]: I used the decouple library as suggested so now all constants and user-config attributes are in settings.py

File: bulkLoadDatasets.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+import csv
+import time
+import traceback
+import re
+import sys
+import pandas
+import requests
+import logging
+
+import configmgrast as config
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py]: this was a mistake by me - I had an example config locally which was not committed that I was testing with called configmgrast, this is now removed

File: bulkLoadDatasets.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+
+def get_file(url):
+    try:
+        logging.info(f'Getting file from {url}')
+        r = requests.get(url)
+        with open(config.MAPPINGS_LOCAL, 'wb') as f:
+            f.write(r.content)
+        return True
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py]: I've improved the exception handling in general to throw some more specific exceptions and sys.exit when it should

File: bulkLoadDatasets.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    parser.add_argument('-v', '--version', action='version', version='%(prog)s 1.0.0')
+    return parser
+
+
+run_pattern = re.compile('^[EDS]RR[0-9]{6,7}$')
+source_pattern = re.compile(config.SOURCE_PATTERN)
+MGX_GET = 'https://wwwdev.ebi.ac.uk/ena/registry/metagenome/api/sequences/{}/datasets'
+MGX_POST = 'https://wwwdev.ebi.ac.uk/ena/registry/metagenome/api/admin/datasets'
+
+
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py]: I added the retry decorators for both the get and post methods for the registry - I set this for any general request exceptions (not including HTTP errors though I'm not sure if this is too broad?

File: bulkLoadDatasets.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        return True
+    except Exception as e:
+        logging.error("Error with request: {0}".format(e))
+        return False
+
+
+def convert_to_tsv(file_location):
+    try:
+        if config.MAPPINGS_FORMAT == 'tsv':
+            return True
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py]: I think I handled the bool originally but removed this - this is now removed.

File: bulkLoadDatasets.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+if __name__ == '__main__':
+    parser = set_parser()
+    args = parser.parse_args()
+    logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
+    try:
+        logging.info(f'Loading datasets into MGX registry for broker: {config.BROKER_ID}')
+        get_file(config.MAPPINGS_DOWNLOAD)
+        convert_to_tsv(config.MAPPINGS_LOCAL)
+
+        with open(config.MAPPINGS_LOCAL) as fd:
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py]: added read mode here

File: bulkLoadDatasets.py [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+                      f'| {file_row[config.SOURCE_ID_COLUMN]}')
+        return None
+
+
+def is_run_accession(insdc_id):
+    return run_pattern.match(insdc_id)
+
+
+def is_source_accession(source_id):
+    return source_pattern.match(source_id)
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py]: I think I copied this from enaBrowserTools but have removed as suggested

File: bulkLoadDatasets.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+            file.to_csv(file_location, sep="\t", index=False)
+            return True
+        else:
+            return False
+    except Exception as e:
+        logging.error("Error with file conversion: {0}".format(e))
+        return False
+
+def convert_into_dataset(file_row):
+    if is_run_accession(file_row[config.INSDC_ID_COLUMN]) and is_source_accession(file_row[config.SOURCE_ID_COLUMN]):
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py]: fixed

File: config.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+MAPPINGS_LOCAL = ''  # Where to download the mappings to on the local file system
+MAPPINGS_FORMAT = ''  # Format of the mappings - options are 'xlsx', 'csv' or 'tsv'
+MAPPINGS_HEADER = True / False  # Whether the mappings have a header line
+SOURCE_ID_COLUMN = 0  # The column index for the analysis IDs
+INSDC_ID_COLUMN = 0  # The column index for the INSDC IDs
+
+# Other required metadata #
+# ----------------------- #
+METHODS = ['other_metadata']  # methods used to derive dataset to sequence mappings
+CONFIDENCE = 'full'  # confidence of the dataset to sequence mappings
```
[the remark this answers is not in the document]
Comment (josieburgin) [reply config.py]: I used the decoupled library as suggested - thanks!

File: bulkLoadDatasets.py [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    return parser
+
+
+run_pattern = re.compile('^[EDS]RR[0-9]{6,7}$')
+source_pattern = re.compile(config.SOURCE_PATTERN)
+MGX_GET = 'https://wwwdev.ebi.ac.uk/ena/registry/metagenome/api/sequences/{}/datasets'
+MGX_POST = 'https://wwwdev.ebi.ac.uk/ena/registry/metagenome/api/admin/datasets'
+
+
+def get_file(url):
```
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply bulkLoadDatasets.py:32]: For docstring I would also recommend to provide description of parameters and outputs like this (example from my code):
def do_additional_checks(self, url, name):
"""
This function will return warning if recommended fields is not present
in record
:param url: schema url for this record
:param name: name of the record
:return: warnings: list object with warning messages
"""
Then using different tools (for example https://www.sphinx-doc.org/en/master/) it's possible to create interactive docs

File: bulkLoadDatasets.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Alexey-ebi) [reply bulkLoadDatasets.py]: I couldn't find retry decorator for **get_file** and **dataset_is_public** functions; it's up to you to decide whether you need it there or not, fell free to resolve this conversation.

File: bulkLoadDatasets.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py:33]: Great thanks - I've added the params/return to the docstrings now

File: bulkLoadDatasets.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (josieburgin) [reply bulkLoadDatasets.py]: I've added retries to those methods too now, thanks

**Pull request merged**